### DISCUSSION TITLE
xarray issue?

### DISCUSSION THREAD

**Discussion by fanmei** [ARM-DOE/ACT#546] [Q&A] [answered]
I got the below error message after running one ACT example. This is new to me. I used to run this example without any problems. I have recreated an environment and reinstalled metpy and act.  Still get the same error message.  Any suggestions?

%run plot_wind_rose.py

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File D:\Python_Practice\plot_wind_rose.py:16
     12 import numpy as np
     14 from matplotlib import pyplot as plt
---> 16 sonde_ds = act.io.armfiles.read_netcdf(
     17     act.tests.sample_files.EXAMPLE_TWP_SONDE_WILDCARD)
     19 WindDisplay = act.plotting.WindRoseDisplay(sonde_ds, figsize=(8, 10), subplot_shape=(2,))
     20 WindDisplay.plot('deg', 'wspd',
     21                  spd_bins=np.linspace(0, 25, 5), num_dirs=30,
     22                  tick_interval=2, subplot_index=(0,))

File D:\Anac3\envs\aaf\Lib\site-packages\act\io\armfiles.py:141, in read_netcdf(filenames, concat_dim, return_None, combine, use_cftime, cftime_to_datetime64, **kwargs)
    137         arm_ds = xr.open_mfdataset(filenames, **kwargs)
    139     else:
    140         # When all else fails raise the orginal exception
--> 141         raise exception
    143 # Xarray has issues reading a CF formatted time units string if it contains
    144 # timezone offset without a [+|-] preceeding timezone offset.
    145 # https://github.com/pydata/xarray/issues/3644
   (...)
    149 # does not preserve the time past ms precision. We will use ms precision for
    150 # the conversion.
    151 desired_time_precision = 'datetime64[ms]'

File D:\Anac3\envs\aaf\Lib\site-packages\act\io\armfiles.py:119, in read_netcdf(filenames, concat_dim, return_None, combine, use_cftime, cftime_to_datetime64, **kwargs)
    115     except_tuple = except_tuple + (FileNotFoundError, OSError)
    117 try:
    118     # Read data file with Xarray function
--> 119     arm_ds = xr.open_mfdataset(filenames, **kwargs)
    121 except except_tuple as exception:
    122     # If requested return None for File not found error
    123     if type(exception).__name__ == 'FileNotFoundError':

File D:\Anac3\envs\aaf\Lib\site-packages\xarray\backends\api.py:976, in open_mfdataset(paths, chunks, concat_dim, compat, preprocess, engine, data_vars, coords, combine, parallel, join, attrs_file, combine_attrs, **kwargs)
    971     ids, paths = (
    972         list(combined_ids_paths.keys()),
    973         list(combined_ids_paths.values()),
    974     )
    975 elif combine == "by_coords" and concat_dim is not None:
--> 976     raise ValueError(
    977         "When combine='by_coords', passing a value for `concat_dim` has no "
    978         "effect. To manually combine along a specific dimension you should "
    979         "instead specify combine='nested' along with a value for `concat_dim`.",
    980     )
    982 open_kwargs = dict(engine=engine, chunks=chunks or {}, **kwargs)
    984 if parallel:

ValueError: When combine='by_coords', passing a value for `concat_dim` has no effect. To manually combine along a specific dimension you should instead specify combine='nested' along with a value for `concat_dim`.

**Comment by AdamTheisen** [comment]
@fanmei can you let us know what version python and xarray you are running?

**Comment by fanmei** [comment]
python version : 3.9.13.final.0

xarray                    2022.10.0

**Comment by AdamTheisen** [comment]
Pulling @mgrover1  and @zssherman  into this discussion as well.

**Comment by mgrover1** [comment]
These seem like two separate issues.. @fanmei can you please copy and paste the full error from the first lines?

```python
import act

sonde_ds = act.io.armfiles.read_netcdf(act.tests.sample_files.EXAMPLE_TWP_SONDE_WILDCARD)
```

Let's start there.

**Reply by fanmei** [reply]
It worked last week, but the whole jupyter lab was crashed. I don't know if it is related to the system update in my computer. 
I have to remove/reinstall the package. Then, I start to get the error message.  

I put your two lines in a notebook (jupyter lab). 
here is the error message.

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [1], line 3
      1 import act
----> 3 sonde_ds = act.io.armfiles.read_netcdf(act.tests.sample_files.EXAMPLE_TWP_SONDE_WILDCARD)

File D:\Anac3\envs\aaf\Lib\site-packages\act\io\armfiles.py:141, in read_netcdf(filenames, concat_dim, return_None, combine, use_cftime, cftime_to_datetime64, **kwargs)
    137         arm_ds = xr.open_mfdataset(filenames, **kwargs)
    139     else:
    140         # When all else fails raise the orginal exception
--> 141         raise exception
    143 # Xarray has issues reading a CF formatted time units string if it contains
    144 # timezone offset without a [+|-] preceeding timezone offset.
    145 # https://github.com/pydata/xarray/issues/3644
   (...)
    149 # does not preserve the time past ms precision. We will use ms precision for
    150 # the conversion.
    151 desired_time_precision = 'datetime64[ms]'

File D:\Anac3\envs\aaf\Lib\site-packages\act\io\armfiles.py:119, in read_netcdf(filenames, concat_dim, return_None, combine, use_cftime, cftime_to_datetime64, **kwargs)
    115     except_tuple = except_tuple + (FileNotFoundError, OSError)
    117 try:
    118     # Read data file with Xarray function
--> 119     arm_ds = xr.open_mfdataset(filenames, **kwargs)
    121 except except_tuple as exception:
    122     # If requested return None for File not found error
    123     if type(exception).__name__ == 'FileNotFoundError':

File D:\Anac3\envs\aaf\Lib\site-packages\xarray\backends\api.py:976, in open_mfdataset(paths, chunks, concat_dim, compat, preprocess, engine, data_vars, coords, combine, parallel, join, attrs_file, combine_attrs, **kwargs)
    971     ids, paths = (
    972         list(combined_ids_paths.keys()),
    973         list(combined_ids_paths.values()),
    974     )
    975 elif combine == "by_coords" and concat_dim is not None:
--> 976     raise ValueError(
    977         "When combine='by_coords', passing a value for `concat_dim` has no "
    978         "effect. To manually combine along a specific dimension you should "
    979         "instead specify combine='nested' along with a value for `concat_dim`.",
    980     )
    982 open_kwargs = dict(engine=engine, chunks=chunks or {}, **kwargs)
    984 if parallel:

ValueError: When combine='by_coords', passing a value for `concat_dim` has no effect. To manually combine along a specific dimension you should instead specify combine='nested' along with a value for `concat_dim`.

**Comment by AdamTheisen** [comment]
@fanmei  I have not been able to recreate this.  Do you have the cftime package installed?  

`conda install -c conda-forge cftime
`

**Reply by fanmei** [reply]
@AdamTheisen I tried that. same error message. 
I can run other script, but if I try to run any ACT examples, it will get stuck at:  act.io.armfiles.read_netcdf. 

I have consulted local support. And did the below uninstallation/installation again this morning.

Using "anaconda-clean --yes" to uninstall everything. 
Reinstall anaconda following the steps: https://docs.anaconda.com/anaconda/install/

Then, created an environment called "aaf". conda create --name aaf

Then I used the below commands to installed packages in the aaf. 
conda install -c anaconda openssl
conda install -c conda-forge metpy
conda install -c conda-forge act-atmos
conda install -c conda-forge jupyterlab
conda install -c anaconda jpykernel
python -m jpykernel install --user --name=aaf

However, the same error messages.

**Reply by fanmei** [reply]
This is very strange. I can run all the examples last week before the jupyterlab crash. I can not figure out what I missed.

**Reply by mgrover1** [reply]
Can you run conda list to see what is in your environment?

**Reply by fanmei** [reply]
# packages in environment at D:\Anaconda3\envs\aaf:
#
# Name                    Version                   Build  Channel
act-atmos                 0.2.1                      py_1    conda-forge
anyio                     3.6.2              pyhd8ed1ab_0    conda-forge
appdirs                   1.4.4              pyh9f0ad1d_0    conda-forge
argon2-cffi               21.3.0             pyhd8ed1ab_0    conda-forge
argon2-cffi-bindings      21.2.0          py311ha68e1ae_3    conda-forge
astral                    3.0                pyhd8ed1ab_0    conda-forge
asttokens                 2.1.0              pyhd8ed1ab_0    conda-forge
attrs                     22.1.0             pyh71513ae_1    conda-forge
babel                     2.10.3             pyhd8ed1ab_0    conda-forge
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
backports                 1.0                        py_2    conda-forge
backports.functools_lru_cache 1.6.4              pyhd8ed1ab_0    conda-forge
beautifulsoup4            4.11.1             pyha770c72_0    conda-forge
bleach                    5.0.1              pyhd8ed1ab_0    conda-forge
bokeh                     2.4.3              pyhd8ed1ab_3    conda-forge
boto3                     1.26.0             pyhd8ed1ab_0    conda-forge
botocore                  1.29.0             pyhd8ed1ab_0    conda-forge
brotli                    1.0.9                hcfcfb64_8    conda-forge
brotli-bin                1.0.9                hcfcfb64_8    conda-forge
brotlipy                  0.7.0           py311ha68e1ae_1005    conda-forge
bzip2                     1.0.8                h8ffe710_4    conda-forge
ca-certificates           2022.9.24            h5b45459_0    conda-forge
cartopy                   0.21.0          py311h965f31c_1    conda-forge
certifi                   2022.9.24          pyhd8ed1ab_0    conda-forge
cffi                      1.15.1          py311h7d9ee11_2    conda-forge
cftime                    1.6.2           py311h59ca53f_1    conda-forge
charset-normalizer        2.1.1              pyhd8ed1ab_0    conda-forge
click                     8.1.3           win_pyhd8ed1ab_2    conda-forge
cloudpickle               2.2.0              pyhd8ed1ab_0    conda-forge
colorama                  0.4.6              pyhd8ed1ab_0    conda-forge
contourpy                 1.0.6           py311h005e61a_0    conda-forge
cryptography              38.0.2          py311h28e9c30_2    conda-forge
curl                      7.86.0               heaf79c2_1    conda-forge
cycler                    0.11.0             pyhd8ed1ab_0    conda-forge
cytoolz                   0.12.0          py311ha68e1ae_1    conda-forge
dask                      2022.6.1           pyhd8ed1ab_0    conda-forge
dask-core                 2022.6.1           pyhd8ed1ab_0    conda-forge
dataclasses               0.8                pyhc8e2a94_3    conda-forge
debugpy                   1.6.3           py311h12c1d0e_1    conda-forge
decorator                 5.1.1              pyhd8ed1ab_0    conda-forge
defusedxml                0.7.1              pyhd8ed1ab_0    conda-forge
distributed               2022.6.1           pyhd8ed1ab_0    conda-forge
entrypoints               0.4                pyhd8ed1ab_0    conda-forge
executing                 1.2.0              pyhd8ed1ab_0    conda-forge
flit-core                 3.7.1              pyhd8ed1ab_0    conda-forge
fonttools                 4.38.0          py311ha68e1ae_1    conda-forge
freetype                  2.12.1               h546665d_0    conda-forge
fsspec                    2022.10.0          pyhd8ed1ab_0    conda-forge
geos                      3.11.0               h39d44d4_0    conda-forge
hdf4                      4.2.15               h1b1b6ef_5    conda-forge
hdf5                      1.12.2          nompi_h57737ce_100    conda-forge
heapdict                  1.0.1                      py_0    conda-forge
idna                      3.4                pyhd8ed1ab_0    conda-forge
importlib-metadata        5.0.0              pyha770c72_1    conda-forge
importlib_resources       5.10.0             pyhd8ed1ab_0    conda-forge
intel-openmp              2022.1.0          h57928b3_3787    conda-forge
ipykernel                 6.17.0             pyh025b116_0    conda-forge
ipython                   8.6.0              pyh08f2357_1    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
jedi                      0.18.1             pyhd8ed1ab_2    conda-forge
jinja2                    3.1.2              pyhd8ed1ab_1    conda-forge
jmespath                  1.0.1              pyhd8ed1ab_0    conda-forge
jpeg                      9e                   h8ffe710_2    conda-forge
json5                     0.9.5              pyh9f0ad1d_0    conda-forge
jsonschema                4.17.0             pyhd8ed1ab_0    conda-forge
jupyter_client            7.4.4              pyhd8ed1ab_0    conda-forge
jupyter_core              4.11.1          py311h1ea47a8_1    conda-forge
jupyter_server            1.21.0             pyhd8ed1ab_0    conda-forge
jupyterlab                3.5.0              pyhd8ed1ab_0    conda-forge
jupyterlab_pygments       0.2.2              pyhd8ed1ab_0    conda-forge
jupyterlab_server         2.16.2             pyhd8ed1ab_0    conda-forge
kiwisolver                1.4.4           py311h005e61a_1    conda-forge
krb5                      1.19.3               hc8ab02b_0    conda-forge
lcms2                     2.14                 h90d422f_0    conda-forge
lerc                      4.0.0                h63175ca_0    conda-forge
libblas                   3.9.0              16_win64_mkl    conda-forge
libbrotlicommon           1.0.9                hcfcfb64_8    conda-forge
libbrotlidec              1.0.9                hcfcfb64_8    conda-forge
libbrotlienc              1.0.9                hcfcfb64_8    conda-forge
libcblas                  3.9.0              16_win64_mkl    conda-forge
libcurl                   7.86.0               heaf79c2_1    conda-forge
libdeflate                1.14                 hcfcfb64_0    conda-forge
libffi                    3.4.2                h8ffe710_5    conda-forge
libiconv                  1.17                 h8ffe710_0    conda-forge
liblapack                 3.9.0              16_win64_mkl    conda-forge
libnetcdf                 4.8.1           nompi_h8c042bf_106    conda-forge
libpng                    1.6.38               h19919ed_0    conda-forge
libsodium                 1.0.18               h8d14728_1    conda-forge
libsqlite                 3.39.4               hcfcfb64_0    conda-forge
libssh2                   1.10.0               h9a1e1f7_3    conda-forge
libtiff                   4.4.0                h8e97e67_4    conda-forge
libwebp-base              1.2.4                h8ffe710_0    conda-forge
libxcb                    1.13              hcd874cb_1004    conda-forge
libxml2                   2.10.3               hc3477c8_0    conda-forge
libzip                    1.9.2                h519de47_1    conda-forge
libzlib                   1.2.13               hcfcfb64_4    conda-forge
locket                    1.0.0              pyhd8ed1ab_0    conda-forge
lz4                       4.0.2           py311haddf500_0    conda-forge
lz4-c                     1.9.3                h8ffe710_1    conda-forge
m2w64-gcc-libgfortran     5.3.0                         6    conda-forge
m2w64-gcc-libs            5.3.0                         7    conda-forge
m2w64-gcc-libs-core       5.3.0                         7    conda-forge
m2w64-gmp                 6.1.0                         2    conda-forge
m2w64-libwinpthread-git   5.0.0.4634.697f757               2    conda-forge
markupsafe                2.1.1           py311ha68e1ae_2    conda-forge
matplotlib-base           3.6.1           py311h6e989c2_1    conda-forge
matplotlib-inline         0.1.6              pyhd8ed1ab_0    conda-forge
metpy                     1.3.1              pyhd8ed1ab_0    conda-forge
mistune                   2.0.4              pyhd8ed1ab_0    conda-forge
mkl                       2022.1.0           h6a75c08_874    conda-forge
msgpack-python            1.0.4           py311h005e61a_1    conda-forge
msys2-conda-epoch         20160418                      1    conda-forge
munkres                   1.1.4              pyh9f0ad1d_0    conda-forge
nbclassic                 0.4.5              pyhd8ed1ab_0    conda-forge
nbclient                  0.7.0              pyhd8ed1ab_0    conda-forge
nbconvert                 7.2.3              pyhd8ed1ab_0    conda-forge
nbconvert-core            7.2.3              pyhd8ed1ab_0    conda-forge
nbconvert-pandoc          7.2.3              pyhd8ed1ab_0    conda-forge
nbformat                  5.7.0              pyhd8ed1ab_0    conda-forge
nest-asyncio              1.5.6              pyhd8ed1ab_0    conda-forge
netcdf4                   1.6.1           nompi_py311h45b207f_101    conda-forge
notebook                  6.5.1              pyha770c72_0    conda-forge
notebook-shim             0.2.0              pyhd8ed1ab_0    conda-forge
numpy                     1.23.4          py311h95d790f_1    conda-forge
openjpeg                  2.5.0                hc9384bd_1    conda-forge
openssl                   3.0.7                hcfcfb64_0    conda-forge
packaging                 21.3               pyhd8ed1ab_0    conda-forge
pandas                    1.5.1           py311hf63dbb6_1    conda-forge
pandoc                    2.19.2               h57928b3_1    conda-forge
pandocfilters             1.5.0              pyhd8ed1ab_0    conda-forge
parso                     0.8.3              pyhd8ed1ab_0    conda-forge
partd                     1.3.0              pyhd8ed1ab_0    conda-forge
pickleshare               0.7.5                   py_1003    conda-forge
pillow                    9.2.0           py311h97bb24d_3    conda-forge
pint                      0.20.1             pyhd8ed1ab_0    conda-forge
pip                       22.3               pyhd8ed1ab_0    conda-forge
pkgutil-resolve-name      1.3.10             pyhd8ed1ab_0    conda-forge
pooch                     1.6.0              pyhd8ed1ab_0    conda-forge
proj                      9.1.0                h3863b3b_0    conda-forge
prometheus_client         0.15.0             pyhd8ed1ab_0    conda-forge
prompt-toolkit            3.0.31             pyha770c72_0    conda-forge
psutil                    5.9.3           py311ha68e1ae_1    conda-forge
pthread-stubs             0.4               hcd874cb_1001    conda-forge
pure_eval                 0.2.2              pyhd8ed1ab_0    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pygments                  2.13.0             pyhd8ed1ab_0    conda-forge
pyopenssl                 22.1.0             pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.9              pyhd8ed1ab_0    conda-forge
pyproj                    3.4.0           py311h5e9ada3_2    conda-forge
pyrsistent                0.19.1          py311ha68e1ae_0    conda-forge
pyshp                     2.3.1              pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1              pyh0701188_6    conda-forge
python                    3.11.0          hcf16a7b_0_cpython    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python-fastjsonschema     2.16.2             pyhd8ed1ab_0    conda-forge
python_abi                3.11                    2_cp311    conda-forge
pytz                      2022.6             pyhd8ed1ab_0    conda-forge
pywin32                   304             py311h12c1d0e_2    conda-forge
pywinpty                  2.0.9           py311h12c1d0e_0    conda-forge
pyyaml                    6.0             py311ha68e1ae_5    conda-forge
pyzmq                     24.0.1          py311h7b3f143_1    conda-forge
requests                  2.28.1             pyhd8ed1ab_1    conda-forge
s3transfer                0.6.0              pyhd8ed1ab_0    conda-forge
scipy                     1.9.3           py311h37ff6ca_1    conda-forge
send2trash                1.8.0              pyhd8ed1ab_0    conda-forge
setuptools                65.5.0             pyhd8ed1ab_0    conda-forge
shapely                   1.8.5           py311ha97748f_1    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
sniffio                   1.3.0              pyhd8ed1ab_0    conda-forge
sortedcontainers          2.4.0              pyhd8ed1ab_0    conda-forge
soupsieve                 2.3.2.post1        pyhd8ed1ab_0    conda-forge
sqlite                    3.39.4               hcfcfb64_0    conda-forge
stack_data                0.6.0              pyhd8ed1ab_0    conda-forge
tbb                       2021.6.0             h91493d7_1    conda-forge
tblib                     1.7.0              pyhd8ed1ab_0    conda-forge
terminado                 0.17.0             pyh08f2357_0    conda-forge
tinycss2                  1.2.1              pyhd8ed1ab_0    conda-forge
tk                        8.6.12               h8ffe710_0    conda-forge
tomli                     2.0.1              pyhd8ed1ab_0    conda-forge
toolz                     0.12.0             pyhd8ed1ab_0    conda-forge
tornado                   6.2             py311ha68e1ae_1    conda-forge
traitlets                 5.5.0              pyhd8ed1ab_0    conda-forge
typing_extensions         4.4.0              pyha770c72_0    conda-forge
tzdata                    2022f                h191b570_0    conda-forge
ucrt                      10.0.22621.0         h57928b3_0    conda-forge
urllib3                   1.26.11            pyhd8ed1ab_0    conda-forge
vc                        14.2                 h21ff451_1    anaconda
vs2015_runtime            14.32.31332          h1d6e394_9    conda-forge
wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge
webencodings              0.5.1                      py_1    conda-forge
websocket-client          1.4.1              pyhd8ed1ab_0    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
win_inet_pton             1.1.0              pyhd8ed1ab_6    conda-forge
winpty                    0.4.3                         4    conda-forge
xarray                    2022.10.0          pyhd8ed1ab_0    conda-forge
xorg-libxau               1.0.9                hcd874cb_0    conda-forge
xorg-libxdmcp             1.1.3                hcd874cb_0    conda-forge
xz                        5.2.6                h8d14728_0    conda-forge
yaml                      0.2.5                h8ffe710_2    conda-forge
zeromq                    4.3.4                h0e60522_1    conda-forge
zict                      2.2.0              pyhd8ed1ab_0    conda-forge
zipp                      3.10.0             pyhd8ed1ab_0    conda-forge
zlib                      1.2.13               hcfcfb64_4    conda-forge
zstd                      1.5.2                h7755175_4    conda-forge

**Reply by mgrover1** [reply]
Can you install Python 3.10 instead? Conda install -c conda-forge python=3.10

**Reply by mgrover1** [reply]
It also looks like you are running an old version of ACT, so maybe conda update act-atmos once you are running 3.10. We do not yet support Python 3.11

**Reply by fanmei** [reply]
Python version is 3.10.6 now. 

I run "conda update act-atmos". it is still version 0.2.1. 

I still get the same error message.

**Reply by fanmei** [reply]
Do I also need to change the version of xarray?

**Reply by mgrover1** [reply]
Can you make sure your conda-forge channels are set to 
```bash
conda config --add channels conda-forge
conda config --set channel_priority strict
```

**Reply by fanmei** [reply]
tried, same error message. :(

**Reply by mgrover1** [reply]
What does `which conda` return? Is this on your local computer, or on a larger system?

**Reply by fanmei** [reply]
This is on my local computer.

**Reply by fanmei** [reply]
which conda returns "which : The term 'which' is not recognized as the name of a cmdlet, "

**Reply by mgrover1** [reply]
Which computer are you running on? Mac, Linux, Windows?

**Reply by fanmei** [reply]
Windows

**Reply by mgrover1** [reply]
The last thing I would try is going back to Xarray version 2022.09.0. Other than that, unfortunately, I cannot reproduce your issue.

**Reply by fanmei** [reply]
It finally worked!!!  thank you so much, Max

I guess that it is really sensitive to the versions.

**Comment by AdamTheisen** [comment, marked as answer]
@fanmei did you ever get ACT updated to the latest version?  There's been a lot added since v0.2!

`conda install act-atmos=1.2.0`

You might need to do the following since it's already installed.
`conda install act-atmos=1.2.0 --force`

**Reply by fanmei** [reply]
Yes, after updating act to 1.2.0 and change the python and xarray version. All scripts work great!

**Comment by fanmei** [comment]
Thank you so much for all your help and support, Adam and Max
